### DISCUSSION TITLE
NetBox SaaS announcement by NS1

### DISCUSSION THREAD

**Discussion by jeremystretch** [netbox-community/netbox#7033] [Announcements]
Hey folks! I wanted to share the news that NS1 is working on a new SaaS offering for NetBox dubbed [NetBox Cloud](https://ns1.com/blog/introducing-netbox-cloud-by-ns1). This is essentially the same open source NetBox application offered as a hosted product, extended for cloud-based management and bundled with professional support. I'll be leading the development internally and hope to be able to share more detail about the upcoming product in the near future.

We decided to launch NetBox Cloud for three primary reasons:

* **Provide commercial support.** Many existing NetBox users have been asking for a paid support option, and other potential users have expressed the need for a professional support agreement to be in place merely to adopt any open source tool. NetBox Cloud fills both needs.

* **Lower the barrier to adoption.** The option of a SaaS product minimizes the time needed for organizations to get up and running with NetBox, especially those teams without the internal resources needed to deploy and manage critical applications. We're confident that offering an expedited deployment path will help grow the overall installation base, and with it, the NetBox community.

* **Commitment to open source.** With NetBox Cloud, NS1 is putting skin in the game. This is a huge milestone for NetBox itself, and represents its recognition as a robust, mature solution for network infrastructure modeling. I'm extremely proud to see this next phase in NetBox's evolution taking shape.

I also want to reassure the community that NetBox itself remains unchanged. Rather than distracting from NetBox development, I expect the introduction of NetBox Cloud to reinforce and reinvigorate the open source project. Advances made in NetBox's development will benefit users of the free and paid products alike.

If NetBox Cloud sounds appealing, be sure to [sign up for early access](https://www.getnetbox.io/). And of course, let me know if you have any questions!

**Comment by candlerb** [comment]
I think this is a great idea, especially if the cloud-hosted and open source versions remain identical - because it means you can have local dev/test instances and be confident that they will work identically when deployed to the cloud.

There is one point I think is necessary to mention, which you've probably already considered, and that's API stability.

It would be a major pain if you've done a load of work to integrate a cloud-based Netbox with other systems, and then either:

1. Overnight it changes, and breaks all all your integrations
2. It doesn't change until you click a button, but sticking to the current version leaves a severe bug or security hole

So I think for a cloud-hosted Netbox to be really attractive, there needs to be a way to keep APIs stable for an extended period of time, especially the REST API (and GraphQL going forward).

One possibility would be to have "LTS" branches of Netbox supported for a certain amount of time.  Those branches would receive fixes for important bugs, and/or very small and safe incremental improvements (e.g. new interface types), or even just security fixes.  The functionality would otherwise not change.  Behind the scenes, Netbox LTS branches might need minor updates to continue to work with newer versions of Python, or Django, or other libraries it depends on.  I expect the minimal-work approach would be to synchronize Netbox branches with Django LTS versions.

The advantage of this approach is that *everything* should continue to work - even custom plugins, scripts and reports (although I don't know whether you'll be allowing users to upload those).  The user would get, say, 3 or 5 years of use out of a given branch of Netbox, and could opt to do the upgrade at a time that makes sense to them, when the new features justify the work required to update the integrations.

The other approach would be to have a strongly versioned API, where each API version is supported for a set amount of time, multiple API versions are exposed simultaneously, and the older APIs are adapted to work with changes to the underlying data models - perhaps by means of database views for backwards compatibility.  This is highly desirable from the user's point of view because it allows individual systems to be updated to use newer versions of the API one by one over time, without a "big bang" from switching Netbox to a new version (incidentally, a big bang which is likely to be irreversible).  It would be a lot more work for Netbox to maintain multiple versions of the REST/GraphQL APIs, and it would be especially difficult to support multiple versions of the Python models for use by scripts/reports etc.  There's also a lot of testing required to ensure that older APIs really don't change. However this is how "grown-up" SaaS APIs like [Salesforce](https://developer.salesforce.com/docs/atlas.en-us.api_rest.meta/api_rest/dome_versions.htm) work.

Finally, a related useful feature for cloud-hosted Netbox would be to be able to clone the production instance to a development instance.  This would allow you to:
1. develop against the API, without risking changes to live data
2. test upgrades to a newer version of Netbox, without risking live service
3. run user training sessions and demos safely
